### ISSUE TITLE
Page up/down and home/end introduce box-with-question-mark-character

### ISSUE THREAD

**Issue by miguelcmedeiros** [superlistapp/super_editor#351]
Steps to reproduce:
* Run super_editor's demo app on macOS
* Place the caret anywhere in a paragraph node
* Press Page up/down or home/end

Expected: whatever macOS usually does when these keys are pressed in the context of text editing (probably needs to be spec'ed out)
Actual: a new character (box-with-question-mark-character) is added at the position where caret is
https://user-images.githubusercontent.com/31278849/142051000-238e3fd3-fe39-4f3f-a28a-f3775e85c96b.mov

Note: on web nothing happens when pressing these keys.

**Comment by bk-one** [comment]
@miguelcmedeiros closing this - I think it's fixed, please re-open if still relevant

**Comment by miguelcmedeiros** [comment]
I can still reproduce it

**Comment by matthew-carroll** [comment]
@miguelcmedeiros @bk-one what should those buttons do? I don't know what the standard caret movement rules are for those keys, or scrolling rules when there is no caret position.

**Comment by bk-one** [comment]
@matthew-carroll I think the names are pretty self-explanatory. However, I also don't think that the implementation is very important. So we might want to split this into various action items:
- [ ] 1. Make sure the keys do not produce question-mark-box characters
- [ ] 2. Implement page-up and page-down to scroll one full viewport length up or down. The caret stays at its current position and is not moved
- [ ]  3. Implement [home] and [end], scroll to the first line or last line of the document. The caret stays at its current position and is not moved.
- [ ] 4. double-check that on caret movement (e.g. cursor keys or entering a character) the viewport is scrolled back to have the caret visible.

**Comment by matthew-carroll** [comment]
@faithoflifedev - would you like to try to implement Page Up, Page Down, Home, and End keyboard handlers for the `SuperEditor` widget?

**Comment by faithoflifedev** [comment]
I'll take a look and let you know if it's something I can handle.

**Comment by faithoflifedev** [comment]
@matthew-carroll I have some initial code that covers step 1.  Currently, when a user presses any "F" key, the question-mark-box will be displayed, so the code covers all "F" keys.

```dart
//document_keyboard_actions.dart
ExecutionInstruction doNothingWhenFKeyPressed({
  required EditContext editContext,
  required RawKeyEvent keyEvent,
}) {
  if (keyEvent.logicalKey.keyId < LogicalKeyboardKey.f1.keyId ||
      keyEvent.logicalKey.keyId > LogicalKeyboardKey.f12.keyId) {
    return ExecutionInstruction.continueExecution;
  }

  return ExecutionInstruction.haltExecution;
}
```

For step 2 and beyond, it looks like things get more complicated.  I'm not yet clear on how the "caret" state would be added to the application

**Comment by matthew-carroll** [comment]
@faithoflifedev - I don't have the specifics off the top of my head, but the `EditContext` should give you access to the `DocumentLayout`, which is probably needed to calculate how far to scroll, and it also provides access to the `DocumentComposer`, which holds the current document selection.

If you believe you need access to other capabilities, please describe the specifics.

**Comment by faithoflifedev** [comment]
@matthew-carroll , it's possible I don't have a good enough understanding of the Document model/workflow.  

My understanding for caret navigation is the user first presses the activation/deactivation key (in a browser that's F7) to put the application into or out of caret navigation mode, after this additional keys are used for the actual navigation (arrow keys, home, end page-up/down, etc).

So for this functionality to be added to super_editor, the first step would be to store the current caret navigation mode (on or off) and have that mode retrievable in an [ExecutionInstruction].

Once the above is in place, then it should be possible to use the objects you mentioned earlier to navigate within the document.

**Comment by faithoflifedev** [comment]
@matthew-carroll , it looks like I don't have a good enough understanding of the functionality required for the use cases for steps 2 and beyond.

Looking back at the original issue, the problem I seem to have is with "Place the caret anywhere in a paragraph node".  I originally understood this as placing the document in "caret navigation" mode as could be done by pressing F7 in a browser like Chrome.  I have since found that some caret navigation features are already available in super_editor, like move by word with Option-ArrowRight.  So I guess I should first ask the question, what does "Place the caret anywhere in a paragraph node" mean and how is it accomplished.

If it turns out that what's really needed for this is to have the keys like end, home, pageDown and pageUp work as the user would expect (instead of inserting question-mark-box), then I think I can handle that by working with the objects you've specified earlier.

**Comment by matthew-carroll** [comment]
I think the concept of caret navigation mode can be ignored. I wasn't even aware that was a thing. Whenever I've used Page Up and Page Down, I've done so directly, without pressing any function keys.

I think the goal right now is to implement the 4 checkboxes that Ben enumerated above.

**Comment by faithoflifedev** [comment]
@matthew-carroll 

I have the sample code below which performs the home/end operations.  

I haven't been able to figure out how to navigate the document without moving the caret (all of the current `defaultKeyboardActions` appear to do so).  Also, I've looked through the `DocumentLayout` methods and none appear to give me a means of finding an `Offset` or `DocumentNode` at the start or end of the users on-screen visible nodes, which would allow me to calculate where to pgUp/pgDown to.

```dart
DocumentNode? destinationNode;
DocumentPosition? documentPosition;

//end key or fn-arrowDn (on Mac)
if (keyEvent.logicalKey == LogicalKeyboardKey.end) {
  destinationNode = editContext.editor.document.nodes.last;

  documentPosition = DocumentPosition(
    nodeId: destinationNode.id,
    nodePosition: destinationNode.endPosition,
  );
  //home key or fn-arrowUp (on Mac)
} else if (keyEvent.logicalKey == LogicalKeyboardKey.home) {
  destinationNode = editContext.editor.document.nodes.first;

  documentPosition = DocumentPosition(
    nodeId: destinationNode.id,
    nodePosition: destinationNode.beginningPosition,
  );
}

if (documentPosition != null) {
  editContext.composer.selection = DocumentSelection.collapsed(position: documentPosition);
}
```

**Comment by matthew-carroll** [comment]
It sounds like the root problem here is that the handlers don't have access to the document's viewport. This leads to a question of API design, which typically requires that we look at a few approaches and the tradeoffs in terms of what we expose publicly. Maybe we can start with the most direct approach as a prototype, and then re-evaluate?

All of the keyboard handlers are called from within the `SuperEditor` widget. Not necessarily at the top level of that tree, but somewhere within the widget subtree. The `SuperEditor` widget also knows about its scrolling viewport, because `SuperEditor` supports auto-scrolling. By default, `SuperEditor` installs its own `Scrollable`, but if there's an ancestor `Scrollable`, then `SuperEditor` controls the ancestor, instead, so that you can place a `SuperEditor` widget within a broader list.

I think what we should do is edit the `EditContext` that's passed to the keyboard handlers. We should try adding whichever artifacts will allow us to make decisions about scrolling. Maybe that's a `ScrollController`, or maybe it's a `ScrollPosition`, or both, or something else. But in theory, the `SuperEditor` widget could provide those artifacts within the `EditContext`, and then any given keyboard handler could tell those artifacts where to scroll. This would allow you to scroll up/down without moving the document selection.

https://github.com/superlistapp/super_editor/blob/main/super_editor/lib/src/core/edit_context.dart

**Comment by faithoflifedev** [comment]
@matthew-carroll , thanks for the suggestion, I believe I have code now that covers all of the original points made by @bk-one.  The following snippet covers the operation of the navigation keys without moving the caret.
```dart
  if (editContext.scrollController == null) {
    return ExecutionInstruction.continueExecution;
  }

  double? jumpTo;

  //end key or fn-arrowDn (on Mac)
  if (keyEvent.logicalKey == LogicalKeyboardKey.end) {
    jumpTo = editContext.scrollController!.position.maxScrollExtent;
    //home key or fn-arrowUp (on Mac)
  } else if (keyEvent.logicalKey == LogicalKeyboardKey.home) {
    jumpTo = editContext.scrollController!.position.minScrollExtent;
    //pgDn key or fn-arrowDn (on Mac)
  } else if (keyEvent.logicalKey == LogicalKeyboardKey.pageDown) {
    final nextJump = editContext.scrollController!.offset + editContext.scrollController!.position.extentInside;

    jumpTo = nextJump < editContext.scrollController!.position.maxScrollExtent
        ? nextJump
        : editContext.scrollController!.position.maxScrollExtent;
    //pgUp key or fn-arrowUp (on Mac)
  } else if (keyEvent.logicalKey == LogicalKeyboardKey.pageUp) {
    final nextJump = editContext.scrollController!.offset - editContext.scrollController!.position.extentInside;

    jumpTo = nextJump > 0 ? nextJump : editContext.scrollController!.position.minScrollExtent;
  }

  if (jumpTo != null) {
    editContext.scrollController?.jumpTo(jumpTo);
  }
```

**Comment by matthew-carroll** [comment]
@faithoflifedev overall the approach looks reasonable. I have some questions, like why the `scrollController` would be nullable, and whether or not your `pageDown` and `pageUp` implementations would actually jump to the end or beginning of the document. But I think it would probably be a good idea to write a suite of failing tests that cover the various boundary conditions for each of these behaviors, and then you can try this implementation and see if those tests pass. We'll need the tests anyway, so we might as well start with them.

**Comment by faithoflifedev** [comment]
@matthew-carroll for`scrollController`, looking through the code examples I see that it's nullable, so I' expect that there is a use-case where the `scrollController` would not be available to the `editContext`.

For `pageDown` and `pageUp`, the idea is to scroll the length of the current viewport or to the scroll extent if the length goes beyond the extent.  UI is not my strong suit so my login might not be rock solid.  I'll put together tests as you've suggested.
